### PR TITLE
OpenTelemetry tracing

### PR DESCRIPTION
**Pull request by mhelleborg**
## Description

Ported Proto.OpenTracing to use System.Diagnostics ActivitySource with  OpenTelemetry conventions and span propagation.

## Purpose
This pull request is a:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/Proto.OpenTelemetry/OpenTelemetryDecorators.cs [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+            => OpenTelemetryMethodsDecorators.Receive(envelope, _receiveActivitySetup, () => base.Receive(envelope));
+    }
+
+    static class OpenTelemetryMethodsDecorators
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void Send(PID target, object message, ActivitySetup sendActivitySetup, Action send)
+        {
+            using var activity =
+                OpenTelemetryHelpers.BuildStartedScope(Activity.Current?.Context ?? default, nameof(Send), message, sendActivitySetup);
```
Comment (marcinbudny) [inline src/Proto.OpenTelemetry/OpenTelemetryDecorators.cs]: How about using activity names like `proto.send {messageType}` etc. This way it is easy to distinguish proto spans from e.g. spans related to message brokers.

File: src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            string verb,
+            object message,
+            ActivitySetup activitySetup,
+            ActivityKind activityKind = ActivityKind.Internal
+        )
+        {
+            var messageType = message?.GetType().Name ?? "Unknown";
+
+            var name = $"{verb} {messageType}";
+            var tags = new[] {new KeyValuePair<string, object?>(ProtoTags.MessageType, messageType)};
```
Comment (marcinbudny) [inline src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs:30]: How about including `net.host.name` tag with value from `Dns.GetHostName()`? Then we know which host the sending / receiving actor is on. Or should this be added to all outgoing traces by some middleware on tracing pipe?

BTW the name `net.host.name` comes from [SemanticConventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/README.md)

File: src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+            var activity = ActivitySource.StartActivity(name, activityKind, parent, tags);
+
+            activitySetup.Invoke(activity, message!);
```
Comment (marcinbudny) [inline src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs]: Check activity for null?

File: src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+using System.Runtime.CompilerServices;
+using OpenTelemetry;
+using OpenTelemetry.Context.Propagation;
+
+namespace Proto.OpenTelemetry
+{
+    static class OpenTelemetryHelpers
+    {
+        private static readonly ActivitySource ActivitySource = new ActivitySource(ProtoTags.ActivityName);
+        private static readonly TextMapPropagator Propagator = new TraceContextPropagator();
```
Comment (marcinbudny) [inline src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs]: `Propagators.DefaultTextMapPropagator`?

It includes baggage as well https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Sdk.cs#L32

File: src/Proto.OpenTelemetry/OpenTelemetryExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
+﻿using System.Diagnostics;
+
+namespace Proto.OpenTelemetry
+{
+    public delegate void ActivitySetup(Activity? activity, object message);
+
+    public static class OpenTelemetryExtensions
```
Comment (marcinbudny) [inline src/Proto.OpenTelemetry/OpenTelemetryExtensions.cs:8]: How about extension that adds source to `TracerProviderBuilder`

```csharp
public static TracerProviderBuilder AddProtoActorInstrumentation(this TracerProviderBuilder builder) 
    => builder.AddSource(ProtoTags.ActivityName);
```

File: src/Proto.OpenTelemetry/ProtoTags.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
+﻿namespace Proto.OpenTelemetry
+{
+    public static class ProtoTags
+    {
+        public const string ActivityName = "Proto.Actor";
```
Comment (marcinbudny) [inline src/Proto.OpenTelemetry/ProtoTags.cs]: `ActivitySourceName`?

File: src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+        public static void DefaultSetupActivity(Activity _, object __)
+        {
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Activity? BuildStartedScope(
```
Comment (marcinbudny) [inline src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs]: `BuildStartedActivity` ?

File: src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs (hunk 2 above)
Comment (mhelleborg) [reply src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs:30]: We already know from the PID which host it is on, so not sure that is needed. I don't want to unnecessarily bloat the traces

File: src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs (hunk 4 above)
Comment (mhelleborg) [reply src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs]: Sure

File: src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs (hunk 3 above)
Comment (mhelleborg) [reply src/Proto.OpenTelemetry/OpenTelemetryHelpers.cs]: ActivitySetup already has activity as nullable. Not sure if there are any point to calling it if it is actually null though

File: src/Proto.OpenTelemetry/OpenTelemetryDecorators.cs (hunk 1 above)
Comment (mhelleborg) [reply src/Proto.OpenTelemetry/OpenTelemetryDecorators.cs]: Certainly, we can make the operation names a bit more namespaced

File: src/Proto.OpenTelemetry/OpenTelemetryExtensions.cs (hunk 5 above)
Comment (mhelleborg) [reply src/Proto.OpenTelemetry/OpenTelemetryExtensions.cs:8]: Good idea, added

File: src/Proto.OpenTelemetry/OpenTelemetryDecorators.cs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            IContext context,
+            ActivitySetup sendActivitySetup,
+            ActivitySetup receiveActivitySetup
+        ) : base(context)
+        {
+            var actorType = context.Actor.GetType().Name;
+            var self = context.Self.ToString();
+            _sendActivitySetup = (activity, message) => {
+                activity?.SetTag(ProtoTags.ActorType, actorType);
+                activity?.SetTag(ProtoTags.ActorPID, self);
```
Comment (mhelleborg) [inline src/Proto.OpenTelemetry/OpenTelemetryDecorators.cs:51]: I'm not 100% sure we need ActorPID here. The PID is already present in the trace, but depending on the span type, it is either  SenderPid or TargetPid. ActorPid is only to normalize this to a single value

**Pull request merged**